File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/ViewModel/MainActivityViewModel.java
package com.shuneault.netrunnerdeckbuilder.ViewModel;

import com.shuneault.netrunnerdeckbuilder.db.CardRepository;
import com.shuneault.netrunnerdeckbuilder.db.IDeckRepository;
import com.shuneault.netrunnerdeckbuilder.game.Card;
import com.shuneault.netrunnerdeckbuilder.game.Deck;
import com.shuneault.netrunnerdeckbuilder.game.Format;
import com.shuneault.netrunnerdeckbuilder.helper.Sorter;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Collections;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import org.jetbrains.annotations.NotNull;

public class MainActivityViewModel extends ViewModel {
    private CardRepository cardRepo;
    private IDeckRepository deckRepo;
    private MutableLiveData<ArrayList<Deck>> mCurrentDecks = new MutableLiveData<>(); // this needs renaming or splitting into runner / corp??

    public MainActivityViewModel(CardRepository cardRepo, IDeckRepository deckRepo){
        this.cardRepo = cardRepo;
        this.deckRepo = deckRepo;
    }

    // Only the selected tab decks
    public MutableLiveData<ArrayList<Deck>> getDecksForSide(String side){
        ArrayList<Deck> decks = new ArrayList<>();

        // Only the selected tab decks
        for (Deck deck: deckRepo.getAllDecks()) {
            if (deck != null && deck.getSide().equals(side)) {
                decks.add(deck);
            }
        }
        // Sort the list
        Collections.sort(decks, new Sorter.DeckSorter());
        mCurrentDecks.setValue(decks);

        return mCurrentDecks;
    }

    public Deck createDeck(String identityCardCode) {
        Card identity = cardRepo.getCard(identityCardCode);
        Format format = cardRepo.getDefaultFormat();
        Deck deck = new Deck(identity, format);
        deckRepo.createDeck(deck);

        return deck;
    }

    public void starDeck(Deck deck, boolean isStarred) {
        deck.setStarred(isStarred);
        deckRepo.saveDeck(deck);
        //todo: sort decks
//        Collections.sort(decks, new Sorter.DeckSorter());
    }

    public void cloneDeck(@NotNull Deck deck) {
        deckRepo.cloneDeck(deck);
    }
}

File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/ViewModel/MainActivityViewModel.kt
package com.shuneault.netrunnerdeckbuilder.ViewModel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import com.shuneault.netrunnerdeckbuilder.db.CardRepository
import com.shuneault.netrunnerdeckbuilder.db.IDeckRepository
import com.shuneault.netrunnerdeckbuilder.game.Card
import com.shuneault.netrunnerdeckbuilder.game.Deck
import com.shuneault.netrunnerdeckbuilder.helper.Sorter.DeckSorter

class MainActivityViewModel(private val cardRepo: CardRepository, private val deckRepo: IDeckRepository) : ViewModel() {
    private var _side: MutableLiveData<String> = MutableLiveData(Card.Side.SIDE_RUNNER)
    var side: String
        get() = _side.value!!
        set(value) {
            _side.value = value
        }

    fun getLiveDecksForSide(side: String): LiveData<List<Deck>> {
        return Transformations.map(deckRepo.allDecksLiveData()) {
            it.sortWith(DeckSorter())
            it.filter { deck -> deck.isSide(side) }
        }
    }

    // Called when the choose identity activity returns when creating a new deck (ListDeckFragment)
    fun createDeck(identityCardCode: String?): Deck {
        val identity = cardRepo.getCard(identityCardCode)
        val format = cardRepo.defaultFormat
        val deck = Deck(identity, format)
        deckRepo.createDeck(deck)
        return deck
    }

    fun starDeck(deck: Deck, isStarred: Boolean) {
        deck.isStarred = isStarred
        deckRepo.saveDeck(deck)
    }

    fun cloneDeck(deck: Deck) {
        deckRepo.cloneDeck(deck)
    }

}